### PR TITLE
Add 'before_create_directory' and 'after_create_directory' hooks

### PR DESCRIPTION
**Pull request by benniledl**
Add 'before_create_directory' and 'after_create_directory' hooks + automated coding standard fix

Trac ticket: https://core.trac.wordpress.org/ticket/44083

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/functions.php [hunk 1]
```diff
@@ -2074,6 +2075,16 @@ function wp_mkdir_p( $target ) {
 		$dir_perms = 0777;
 	}
 
+	/**
+	 * Fires before the directory is created
```
Comment (costdev) [inline src/wp-includes/functions.php]:
```suggestion
	 * Fires before the directory is created.
```
- Docblock subjects should end with a period.

File: src/wp-includes/functions.php [hunk 2]
```diff
@@ -2074,6 +2075,16 @@ function wp_mkdir_p( $target ) {
...
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string $target Full path to attempt to create
+	 * @param int $dir_perms Directory permissions
```
Comment (costdev) [inline src/wp-includes/functions.php]:
```suggestion
	 * @param string $target    Full path to attempt to create.
	 * @param int    $dir_perms Directory permissions.
```
- The types and descriptions used in `@param` annotations should be aligned.
- Parameter descriptions should end with a period.

File: src/wp-includes/functions.php [hunk 3]
```diff
@@ -2087,6 +2098,16 @@ function wp_mkdir_p( $target ) {
 			}
 		}
 
+		/**
+		 * Fires after the directory is created and the permissions are set.
+		 *
+		 * @since 6.4.0
+		 *
+		 * @param string $target Full path to the created directory.
+		 * @param int $dir_perms Directory permissions
```
Comment (costdev) [inline src/wp-includes/functions.php]:
```suggestion
		 * @param string $target    Full path to the created directory.
		 * @param int    $dir_perms Directory permissions.
```
- The types and descriptions used in `@param` annotations should be aligned.
- Parameter descriptions should end with a period.

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for the PR @benni1516 and for your first code contribution to WordPress Core!

I've left some suggested changes to help bring the documentation in line with the project's standards and conventions.

File: src/wp-includes/functions.php [hunk 4]
```diff
@@ -2074,6 +2075,16 @@ function wp_mkdir_p( $target ) {
...
 
+	/**
+	 * Fires before the directory is created.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string $target    Full path to attempt to create.
+	 * @param int    $dir_perms Directory permissions.
+	 */
+	do_action( 'before_create_directory', $target, $dir_perms );
```
Comment (mukeshpanchal27) [inline src/wp-includes/functions.php]: Could you add unit test for new action?

Comment (benniledl) [reply src/wp-includes/functions.php]: Sorry I am not familiar with unit tests, there was one at https://core.trac.wordpress.org/attachment/ticket/44083/44083.5.diff but I wasn't sure if I could adopt that one 1/1
Is the old unit test still good, can I add this one?

Comment (costdev) [reply src/wp-includes/functions.php]: It looks like that older unit test may be doing too much. I can write some unit tests for this and send them through to you @benni1516 🙂

Comment (mukeshpanchal27) [reply src/wp-includes/functions.php]: Thanks @costdev 🙌

**Review by costdev — APPROVED** [review]
Thanks for the updates @benni1516! This looks great to me! 👍

File: tests/phpunit/tests/filesystem/wpMkdirP.php [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	 * @dataProvider data_actions
+	 *
+	 * @param string $hook_name The name of the action hook.
+	 */
+	public function test_wp_mkdir_p_should_not_fire_action_when_the_target_contains_path_traversal( $hook_name ) {
+		$action = new MockAction();
+		add_action( $hook_name, array( $action, 'action' ) );
+
+		// Force a failure by not using `realpath()`.
+		$target = DIR_TESTDATA . "/$hook_name/";
```
Comment (costdev) [inline tests/phpunit/tests/filesystem/wpMkdirP.php]:
```suggestion
		$target = DIR_TESTDATA . "/test_wp_mkdir_p/$hook_name/";
```
Woops, looks like I forgot to add the `/test_wp_mkdir_p` parent here. Doing so means that even if something strange is created due to a bug introduced to `wp_mkdir_p()`, the `tear_down()` method will clean it up.

File: tests/phpunit/tests/filesystem/wpMkdirP.php [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+	 * @dataProvider data_actions
+	 *
+	 * @param string $hook_name The name of the action hook.
+	 */
+	public function test_wp_mkdir_p_should_not_fire_action_when_the_target_contains_path_traversal_with_directory_separator( $hook_name ) {
+		$action = new MockAction();
+		add_action( $hook_name, array( $action, 'action' ) );
+
+		// Force a failure.
+		$target = str_replace( '../', '..' . DIRECTORY_SEPARATOR, DIR_TESTDATA ) . "/$hook_name";
```
Comment (costdev) [inline tests/phpunit/tests/filesystem/wpMkdirP.php]:
```suggestion
		$target = str_replace( '../', '..' . DIRECTORY_SEPARATOR, DIR_TESTDATA ) . "/test_wp_mkdir_p/$hook_name";
```
Same as before: Adding parent.

**Review by costdev — CHANGES_REQUESTED** [review]
Hey @benni1516! I just spotted a couple of tiny adjustments I'd like to make to the tests here for improved stability. When you get a chance could you commit these two suggestions? Thanks!

**Comment by benniledl** [comment]
thanks for the adjustments @costdev !

**Review by costdev — APPROVED** [review]
Thanks @benni1516!

File: tests/phpunit/tests/filesystem/wpMkdirP.php [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+		$action = new MockAction();
+		add_action( 'create_directory_failed', array( $action, 'action' ) );
+
+		add_action(
+			'before_create_directory',
+			function( $target ) {
+				/*
+					* Force a failure by creating a file of the same name
+					* just before `mkdir()` runs.
+					*/
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/filesystem/wpMkdirP.php]:
```suggestion
				/*
				 * Force a failure by creating a file of the same name
				 * just before `mkdir()` runs.
				 */
```

correct docblock indentation

File: src/wp-includes/functions.php [hunk 8]
```diff
@@ -2087,9 +2097,29 @@ function wp_mkdir_p( $target ) {
...
 	}
 
+	/**
+	 * Fires when directory creation fails
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string    $target     Full path to the directory that failed to create
+	 * @param int       $dir_perms  Directory permissions.
+	 */
```
Comment (costdev) [inline src/wp-includes/functions.php]:
```suggestion
	/**
	 * Fires when directory creation fails.
	 *
	 * @since 6.4.0
	 *
	 * @param string $target    Full path to the directory that failed to create
	 * @param int    $dir_perms Directory permissions.
	 */
```
Just adding a `.` and some alignment fixes.

File: src/wp-includes/functions.php [hunk 9]
```diff
@@ -4822,7 +4852,7 @@ function wp_parse_args( $args, $defaults = array() ) {
 	if ( is_object( $args ) ) {
 		$parsed_args = get_object_vars( $args );
 	} elseif ( is_array( $args ) ) {
-		$parsed_args =& $args;
+		$parsed_args = & $args;
```
Comment (costdev) [inline src/wp-includes/functions.php]:
```suggestion
		$parsed_args =& $args;
```
This was an unrelated change that PHPCBF seems to have made. :slightly_smiling_face:

File: tests/phpunit/tests/filesystem/wpMkdirP.php [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
```
Comment (costdev) [inline tests/phpunit/tests/filesystem/wpMkdirP.php]:
```suggestion
				/*
				 * Force a failure by creating a file of the same name
				 * just before `mkdir()` runs.
				 */
```
Just a minor alignment fix.

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for the updates @benni1516! I've left a couple of changes in this review :slightly_smiling_face:

File: src/wp-includes/functions.php [hunk 11]
```diff
@@ -2087,9 +2097,29 @@ function wp_mkdir_p( $target ) {
...
+
 		return true;
 	}
 
+	/**
+	 * Fires when directory creation fails.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string $target    Full path to the directory that failed to create
```
Comment (costdev) [inline src/wp-includes/functions.php]:
```suggestion
	 * @param string $target    Full path to the directory that failed to create.
```

**Review by costdev — APPROVED** [review]
Thanks for updates @benni1516! This looks good to me now 🙂
